# Project 7 -- Rafee Adnan

**TA Help:** 
    
**Collaboration:** 

## Question 1

In [1]:
options(repr.matrix.max.cols=25, repr.matrix.max.rows=200)

In [2]:
options(jupyter.rich_display = F)

In [3]:
library(data.table)

titles <- data.frame(fread("/anvil/projects/tdm/data/movies_and_tv/titles.csv"))
episodes <- data.frame(fread("/anvil/projects/tdm/data/movies_and_tv/episodes.csv"))
people <- data.frame(fread("/anvil/projects/tdm/data/movies_and_tv/people.csv"))
ratings <- data.frame(fread("/anvil/projects/tdm/data/movies_and_tv/ratings.csv"))

In [4]:
head(titles)

  title_id  type  primary_title          original_title         is_adult
1 tt0000001 short Carmencita             Carmencita             0       
2 tt0000002 short Le clown et ses chiens Le clown et ses chiens 0       
3 tt0000003 short Pauvre Pierrot         Pauvre Pierrot         0       
4 tt0000004 short Un bon bock            Un bon bock            0       
5 tt0000005 short Blacksmith Scene       Blacksmith Scene       0       
6 tt0000006 short Chinese Opium Den      Chinese Opium Den      0       
  premiered ended runtime_minutes genres                  
1 1894      NA     1              Documentary,Short       
2 1892      NA     5              Animation,Short         
3 1892      NA     4              Animation,Comedy,Romance
4 1892      NA    12              Animation,Short         
5 1893      NA     1              Comedy,Short            
6 1894      NA     1              Short                   

In [5]:
diffGenres <- unlist(strsplit(unique(titles$genres), ","))

In [6]:
unique(diffGenres)

 [1] "Documentary" "Short"       "Animation"   "Comedy"      "Romance"    
 [6] "Sport"       "News"        "Drama"       "Fantasy"     "Horror"     
[11] "Biography"   "Music"       "War"         "Crime"       "Western"    
[16] "Family"      "Adventure"   "Action"      "History"     "Mystery"    
[21] "\\N"         "Sci-Fi"      "Musical"     "Thriller"    "Film-Noir"  
[26] "Talk-Show"   "Game-Show"   "Reality-TV"  "Adult"      

In [7]:
tail(sort(table(titles[grepl("Comedy", titles$genres), ]$premiered)))


 2013  2014  2018  2015  2017  2016 
71963 77034 78393 80199 83916 84227 

a) shown above

b) 2016 was the year with the most comedies debuted with 84,227.

## Question 2

In [8]:
head(episodes)

  episode_title_id show_title_id season_number episode_number
1 tt0041951        tt0041038      1            9             
2 tt0042816        tt0989125      1            17            
3 tt0042889        tt0989125     NA            \\N           
4 tt0043426        tt0040051      3            42            
5 tt0043631        tt0989125      2            16            
6 tt0043693        tt0989125      2            8             

In [9]:
dim(episodes)

[1] 5900030       4

In [10]:
dim(titles)

[1] 8064259       9

In [13]:
epiTitles <- merge(episodes, titles, by.x = "episode_title_id", by.y = "title_id")

In [14]:
dim(epiTitles)

[1] 5897173      12

In [15]:
head(epiTitles)

  episode_title_id show_title_id season_number episode_number type     
1 tt0041951        tt0041038      1            9              tvEpisode
2 tt0042816        tt0989125      1            17             tvEpisode
3 tt0042889        tt0989125     NA            \\N            tvEpisode
4 tt0043426        tt0040051      3            42             tvEpisode
5 tt0043631        tt0989125      2            16             tvEpisode
6 tt0043693        tt0989125      2            8              tvEpisode
  primary_title                     original_title                    is_adult
1 The Tenderfeet                    The Tenderfeet                    0       
2 Othello                           Othello                           0       
3 The Tragedy of King Richard II/II The Tragedy of King Richard II/II 0       
4 Coriolanus                        Coriolanus                        0       
5 The Life of King Henry V          The Life of King Henry V          0       
6 Julius Caesar       

The merge function creates a new dataframe that includes only the rows that have an equal title id and episode title id combines the information of the number of rows.

## Question 3

In [16]:
noTitleFriendsDF <- epiTitles[epiTitles$show_title_id == "tt0108778", ]

In [17]:
head(noTitleFriendsDF)

      episode_title_id show_title_id season_number episode_number type     
93207 tt0583431        tt0108778      7            8              tvEpisode
93208 tt0583432        tt0108778     10            9              tvEpisode
93209 tt0583433        tt0108778     10            17             tvEpisode
93210 tt0583435        tt0108778      8            1              tvEpisode
93211 tt0583436        tt0108778     10            1              tvEpisode
93212 tt0583437        tt0108778      5            1              tvEpisode
      primary_title                           
93207 The One Where Chandler Doesn't Like Dogs
93208 The One with the Birth Mother           
93209 The Last One                            
93210 The One After I Do                      
93211 The One After Joey and Rachel Kiss      
93212 The One After Ross Says Rachel          
      original_title                           is_adult premiered ended
93207 The One Where Chandler Doesn't Like Dogs 0        2000      N

In [18]:
friendsDFnoRating <- merge(noTitleFriendsDF, titles, by.x = "show_title_id", by.y = "title_id")

In [19]:
friendsDFrating <- merge(friendsDFnoRating, ratings, by.x = "episode_title_id", by.y = "title_id")

In [33]:
head(friendsDFrating)

  episode_title_id show_title_id season_number episode_number type.x   
1 tt0583431        tt0108778      7            8              tvEpisode
2 tt0583432        tt0108778     10            9              tvEpisode
3 tt0583433        tt0108778     10            17             tvEpisode
4 tt0583435        tt0108778      8            1              tvEpisode
5 tt0583436        tt0108778     10            1              tvEpisode
6 tt0583437        tt0108778      5            1              tvEpisode
  primary_title.x                         
1 The One Where Chandler Doesn't Like Dogs
2 The One with the Birth Mother           
3 The Last One                            
4 The One After I Do                      
5 The One After Joey and Rachel Kiss      
6 The One After Ross Says Rachel          
  original_title.x                         is_adult.x premiered.x ended.x
1 The One Where Chandler Doesn't Like Dogs 0          2000        NA     
2 The One with the Birth Mother            0   

In [20]:
friendsDF <- merge(friendsDFrating, ratings, by.x = "show_title_id", by.y = "title_id")[ , c("primary_title.x", "primary_title.y", "rating.x", "rating.y")]

In [21]:
head(friendsDF)

  primary_title.x                          primary_title.y rating.x rating.y
1 The One Where Chandler Doesn't Like Dogs Friends         8.2      8.9     
2 The One with the Birth Mother            Friends         8.6      8.9     
3 The Last One                             Friends         9.7      8.9     
4 The One After I Do                       Friends         8.7      8.9     
5 The One After Joey and Rachel Kiss       Friends         8.5      8.9     
6 The One After Ross Says Rachel           Friends         8.9      8.9     

In [22]:
names(friendsDF) <- c("episode_title", "show_title", "episode_rating", "show_rating")

In [23]:
head(friendsDF)

  episode_title                            show_title episode_rating
1 The One Where Chandler Doesn't Like Dogs Friends    8.2           
2 The One with the Birth Mother            Friends    8.6           
3 The Last One                             Friends    9.7           
4 The One After I Do                       Friends    8.7           
5 The One After Joey and Rachel Kiss       Friends    8.5           
6 The One After Ross Says Rachel           Friends    8.9           
  show_rating
1 8.9        
2 8.9        
3 8.9        
4 8.9        
5 8.9        
6 8.9        

In [32]:
tail(friendsDF[order(friendsDF$episode_rating), ])

    episode_title                     show_title episode_rating show_rating
225 The One with the Videotape        Friends    9.3            8.9        
235 The One with the Proposal         Friends    9.3            8.9        
202 The One with the Prom Video       Friends    9.4            8.9        
167 The One with the Embryos          Friends    9.5            8.9        
3   The Last One                      Friends    9.7            8.9        
21  The One Where Everybody Finds Out Friends    9.7            8.9        

a) shown above

## Question 4

In [34]:
friendsDF <- merge(friendsDFrating, ratings, by.x = "show_title_id", by.y = "title_id")[ , c("primary_title.x", "primary_title.y", "rating.x", "rating.y", "season_number")]

In [41]:
names(friendsDF) <- c("episode_title", "show_title", "episode_rating", "show_rating", "season_number")

In [42]:
head(friendsDF)

  episode_title                            show_title episode_rating
1 The One Where Chandler Doesn't Like Dogs Friends    8.2           
2 The One with the Birth Mother            Friends    8.6           
3 The Last One                             Friends    9.7           
4 The One After I Do                       Friends    8.7           
5 The One After Joey and Rachel Kiss       Friends    8.5           
6 The One After Ross Says Rachel           Friends    8.9           
  show_rating season_number
1 8.9          7           
2 8.9         10           
3 8.9         10           
4 8.9          8           
5 8.9         10           
6 8.9          5           

In [47]:
sznFiveGood <- friendsDF[(friendsDF$episode_rating > 9) & (friendsDF$season_number == 5), ]

In [50]:
print(sznFiveGood)

                         episode_title show_title episode_rating show_rating
21   The One Where Everybody Finds Out    Friends            9.7         8.9
78    The One with All the Resolutions    Friends            9.1         8.9
130       The One with Ross's Sandwich    Friends            9.1         8.9
219 The One with All the Thanksgivings    Friends            9.2         8.9
233           The One in Vegas: Part 2    Friends            9.1         8.9
    season_number
21              5
78              5
130             5
219             5
233             5


a) shown above

## Question 5

In [53]:
results <- friendsDF

In [56]:
season_number = 6

In [58]:
results[results$episode_rating > 9 & results$season_number == season_number,]

    episode_title               show_title episode_rating show_rating
51  The One Where Ross Got High Friends    9.2            8.9        
137 The One with Unagi          Friends    9.1            8.9        
235 The One with the Proposal   Friends    9.3            8.9        
    season_number
51  6            
137 6            
235 6            

In [61]:
subset(results, episode_rating > 9 & season_number == season_number)

    episode_title                                      show_title
3   The Last One                                       Friends   
21  The One Where Everybody Finds Out                  Friends   
51  The One Where Ross Got High                        Friends   
70  The One with the Morning After                     Friends   
78  The One with All the Resolutions                   Friends   
82  The One with Chandler in a Box                     Friends   
99  The One with Monica and Chandler's Wedding: Part 2 Friends   
130 The One with Ross's Sandwich                       Friends   
137 The One with Unagi                                 Friends   
167 The One with the Embryos                           Friends   
173 The One with the Flashback                         Friends   
187 The One with the Jellyfish                         Friends   
202 The One with the Prom Video                        Friends   
205 The One with the Red Sweater                       Friends   
209 The On

In [62]:
subset(results, episode_rating > 9 & season_number == 6)

    episode_title               show_title episode_rating show_rating
51  The One Where Ross Got High Friends    9.2            8.9        
137 The One with Unagi          Friends    9.1            8.9        
235 The One with the Proposal   Friends    9.3            8.9        
    season_number
51  6            
137 6            
235 6            

Subset gets a different value than regular indexing because it uses non standard evaluation. Rather than evalutating, subset creates a temporary dataframe in which it will substitute some of the variables with different values. If you look at the function above, if you use 6 rather than season_number = 6, it displays the correct episodes. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.